### PR TITLE
Redirect standard output to devtools server and logfile

### LINKED ISSUE #393: Crash when TEXTUAL_DEVTOOLS=1 but no log file supplied
Some of the new devtools code assumes there is a logfile available, but of course the user doesn't need to supplied one. The code should be updated to only write to the logfile if it is available.
```
Traceback (most recent call last):
  File "/Users/darrenburns/.pyenv/versions/3.10.0/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/darrenburns/.pyenv/versions/3.10.0/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/darrenburns/code/textual/sandbox/buttons.py", line 15, in <module>
    ButtonsApp.run(css_file="buttons.scss")
  File "/Users/darrenburns/code/textual/src/textual/app.py", line 304, in run
    asyncio.run(run_app())
  File "/Users/darrenburns/.pyenv/versions/3.10.0/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/darrenburns/.pyenv/versions/3.10.0/lib/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/Users/darrenburns/code/textual/src/textual/app.py", line 290, in run_app
    await app.process_messages()
  File "/Users/darrenburns/code/textual/src/textual/app.py", line 475, in process_messages
    self.log_file.write(
AttributeError: 'NoneType' object has no attribute 'write'
```

### PR DESCRIPTION
**Pull request by darrenburns**
Data printed using the builtin `print` will be sent over to the devtools, if you're connected to them.

It will also be written to the log file, if available.

The behaviour of the file-like object that redirects this output mirrors the behaviour of stdout (line buffering).

There's some additional logic in here to handle calls like `print(1, 2, 3)`. A call to print with multiple arguments results in Python calling the redirectors `write` method 3 times. We batch these calls into 1 since they're made from the same `print` statement call. This means in the devtools you only see 1 timestamp/log entry for a single print function call (as expected).

Closes #393 
Closes #385

### CODE REVIEW

File: src/textual/app.py [hunk 1]
```diff
@@ -502,11 +502,17 @@ async def process_messages(self) -> None:
                 mount_event = events.Mount(sender=self)
                 await self.dispatch_message(mount_event)
 
-                self.console = Console()
+                self.console = Console(file=sys.__stdout__)
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/app.py:505]: Probably, but at least this ensures that we're not using any kind of wrapped file.

**Comment by willmcgugan** [comment]
@darrenburns Sorry a created a merge conflict for you 😈

**Review by willmcgugan — CHANGES_REQUESTED** [review]

File: src/textual/devtools/client.py [hunk 2]
```diff
@@ -171,16 +189,17 @@ def is_connected(self) -> bool:
             return False
         return not (self.session.closed or self.websocket.closed)
 
-    def log(self, *objects: Any, path: str = "", lineno: int = 0) -> None:
+    def log(self, log: DevtoolsLog):
```
[the remark this answers is not in the document]
Comment (tusharsadhwani) [reply src/textual/devtools/client.py:192]: I think it's about time you setup mypy on CI. pre-commit.ci perhaps?

[the remark this answers is not in the document]
Comment (darrenburns) [reply src/textual/devtools/client.py:192]: Good catch, re-added.

@tusharsadhwani I ran `mypy` but it doesn't complain about this by default. We will add `mypy` eventually, but the code is in flux and it won't pass typechecking at the moment. We'll get there eventually :)

**Comment by darrenburns** [comment]
Now closes #393 too

**Review by willmcgugan — APPROVED** [review]
LGTM Nicely done with the line buffering.

**Pull request merged**